server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/VariableWidthHistogramAggregator.java

```diff
@@ -0,0 +1,584 @@
+         * Creates a new cluster with  <code>value</code> and appends it to the cluster arrays
+         */
+        private void createAndAppendNewCluster(double value){
+            // Ensure there is space for the cluster
+            clusterMaxes = bigArrays.grow(clusterMaxes, numClusters + 1); //  + 1 because indexing starts at 0
+            clusterMins = bigArrays.grow(clusterMins, numClusters + 1);
+            clusterCentroids = bigArrays.grow(clusterCentroids, numClusters + 1);
+            clusterSizes = bigArrays.grow(clusterSizes, numClusters + 1);
+
+            // Initialize the cluster at the end of the array
+            clusterMaxes.set(numClusters, value);
+            clusterMins.set(numClusters, value);
+            clusterCentroids.set(numClusters, value);
+            clusterSizes.set(numClusters, 1);
+
+            numClusters += 1;
+        }
+
+        /**
+         * Move the last cluster to position <code>idx</code>
+         * This is expensive because a merge map of size <code>numClusters</code> is created, so don't call this method too often
+         *
+         * TODO: Make this more efficient
+         */
+        private void moveLastCluster(int index){
+            if(index != numClusters - 1) {
+
+                // Move the cluster metadata
+                double holdMax = clusterMaxes.get(numClusters-1);
+                double holdMin = clusterMins.get(numClusters-1);
+                double holdCentroid = clusterCentroids.get(numClusters-1);
+                double holdSize = clusterSizes.get(numClusters-1);
+                for (int i = numClusters - 1; i > index; i--) {
+                    // The clusters in range {index ... numClusters - 1} move up 1 index to make room for the new cluster
+                    clusterMaxes.set(i, clusterMaxes.get(i-1));
+                    clusterMins.set(i, clusterMins.get(i-1));
+                    clusterCentroids.set(i, clusterCentroids.get(i-1));
+                    clusterSizes.set(i, clusterSizes.get(i-1));
+                }
+                clusterMaxes.set(index, holdMax);
+                clusterMins.set(index, holdMin);
+                clusterCentroids.set(index, holdCentroid);
+                clusterSizes.set(index, holdSize);
+
+                // Move the underlying buckets
+                long[] mergeMap = new long[numClusters];
+                for (int i = 0; i < index; i++) {
+                    // The clusters in range {0 ... idx - 1} don't move
+                    mergeMap[i] = i;
+                }
+                for (int i = index; i < numClusters - 1; i++) {
+                    // The clusters in range {index ... numClusters - 1} shift up
+                    mergeMap[i] = i + 1;
+                }
+                // Finally, the new cluster moves to index
+                mergeMap[numClusters - 1] = index;
+
+                // TODO: Create a moveLastCluster() method that is like mergeBuckets but doesn't require a merge map
```

Comment:
You've done this TODO.

Reply:
I actually haven't done this, but I agree that this comment was a bit confusing. I've updated it so that what I'm referring to is clearer.